test/end2end_test.go

```diff
@@ -7416,8 +7419,23 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
 					}
 					return
 				}
-				if hframe, ok := frame.(*http2.HeadersFrame); ok {
-					sid = hframe.Header().StreamID
+				sid = 0
+				switch fr := frame.(type) {
+				case *http2.HeadersFrame:
+					// Respond after this if we are not waiting for an end
+					// stream or if this frame ends it.
+					if !s.waitForEndStream || fr.StreamEnded() {
+						sid = fr.Header().StreamID
+					}
+
+				case *http2.DataFrame:
+					// Respond after this if we were waiting for an end stream
```

Comment:
Explain this more?
`If we were not waiting for an end stream, this RPC was handled already when the header frame was received, and there's no need to do it again`.

Reply:
Done!